src/coreclr/System.Private.CoreLib/src/System/Runtime/DependentHandle.cs

```diff
@@ -81,7 +81,7 @@ public DependentHandle(object? target, object? dependent)
         /// <exception cref="InvalidOperationException">
         /// Thrown if <see cref="IsAllocated"/> is <see langword="false"/> or if the input value is not <see langword="null"/>.</exception>
         /// <remarks>This property is thread-safe.</remarks>
-        public object? Target
+        public readonly object? Target
```

Comment:
This property has a setter. Marking it `readonly` is very odd.

Reply:
I'm going to reply to this since it's the first comment, but it applies to all others below as well, since it's the same logic. The `readonly` modifier here simply means "this member is not mutating any instance state". Which _is_ true: none of these methods it mutating the `GCHandle` instance (including all the setter accessors as well). All they're doing is just mutating some _other_ property somewhere else, which is fine. I don't see how this would be any different than, say, some value type "builder type" that exposes some `Write`/`Append` method which is also marked `readonly` — as long as that method is, say, appending an item to a wrapped `List<T>` object (or something like that), the method is in fact not touching any instance state. This is just about avoid defensive copies from the C# compiler, but `readonly` per se doesn't mean you can make assumptions on whether a member is mutating _state_ in general. Does that make sense?

Reply:
I'm also hesitant to mark the setter as readonly. Removing `readonly` is paramount to a breaking change, which means we're promising here that the setter won't ever mutate the instance state of this struct. While that's the case today, making this readonly is exposing an implementation detail.

Reply:
Let me call out the two phrases in Steve's response that reflect how I reviewed the pattern in this PR.

> mark the setter as readonly

Setters, by definition, mutate. Marking it `readonly` is a code smell and begs the following question: "if this is 'read only' what exactly is this doing?" 

> making this readonly is exposing an implementation detail

This relates to the former statement. Again, setters mutate. There is nothing more to say. If this one doesn't that is interesting and likely also because of some subtle detail that should not be assumed to be the best way or the concern of the caller. This is particularly true in the NAOT world where more and more runtime components are moving to managed code and where expressing implementation details at the managed boundary is a recipe for future pain.

Reply:
If setters being readonly are too odd, maybe I could just mark the getters as so? I'm still not entirely clear I follow the "setters by definition mutate" argument, given that the point here is specifically about indicating that the instance state doesn't mutate, and not a out _any_ state mutating (see my previous example about array builder types), but regardless, I agree it's not really worth spending too much time over. Perfectly happy to update this PR to just mark the getters, or to abandon it if that's the preferred direction.

Reply:
> maybe I could just mark the getters as so?

That is a reasonable expectation and one I would be interested in seeing.

> I'm still not entirely clear I follow the "setters by definition mutate" argument, given that the point here is specifically about indicating that the instance state doesn't mutate, and not a out any state mutating

That is an implementation detail. We aren't expressing how this particular setter is implemented as it is odd, setters set a value on the instance as a concept. In this case the setter does some other hidden behavior but the vast majority of the time setters change state.

Reply:
Also, please please please... I can't say this enough... please respond to https://github.com/dotnet/runtime/pull/93178#issuecomment-1752141194. Too many times we are sent "optimizations" and they either aren't or have zero measurable impact. Let's make sure we are seeing improvements in some way.

Reply:
I agree that marking it on the `setter` is confusing and likely non-beneficial. It happening to only mutate indirect or static state doesn't mean it will always do that or that all possible implementations may do that.

The `getter` we should really just mark as `readonly` though. It's essentially a wrapper around a trivial field read and will likely never need to mutate. So it makes everyone's lives better by us doing the "right thing" and annotating it.

Reply:
> It's essentially a wrapper around a trivial field read and will likely never need to mutate

If we ever introduce GC mode with read barriers, the getter is going to mutate auxiliary state just like the setter does.

GCHandle is a wrapper around IntPtr that is readonly except for the `Free` method. This implementation detail is implicitly exposed already. Changing it would be a major breaking change. I expect that there is a ton of code out there that depends on `GCHandle.Target` setter to not mutate the GCHandle instance itself and only mutate the auxiliary state.